core/src/main/java/org/infinispan/interceptors/impl/BaseStateTransferInterceptor.java

```diff
@@ -125,8 +126,40 @@ protected final void updateTopologyId(TopologyAffectedCommand command) throws In
       // Without this, we could retry the command too fast and we could get the OutdatedTopologyException again.
       int newTopologyId = Math.max(currentTopologyId(), commandTopologyId + 1);
       command.setTopologyId(newTopologyId);
-      waitForTransactionData(newTopologyId);
-      return visitGetKeysInGroupCommand(context, command);
+      return retryCommandWithTransactionData(newTopologyId, context, command, (ctx, cmd) -> visitGetKeysInGroupCommand(ctx, cmd));
+
+   }
+
+   protected <T extends VisitableCommand> CompletableFuture<Void> retryCommandWithTransactionData(int topologyId,
+        InvocationContext ctx, T command, RetryHandler<T> handler) throws Throwable {
+      CompletableFuture<Void> future = stateTransferLock.transactionDataFuture(topologyId);
+      if (future == null) {
+         getLog().tracef("Retrying command %s for topology %d", command, topologyId);
+         return handler.handle(ctx, command);
+      } else {
+         CompletableFuture<Void> retryFuture = future.thenCompose(ignored -> {
+            try {
+               getLog().tracef("Retrying command %s for topology %d", command, topologyId);
+               return handler.handle(ctx, command);
+            } catch (Throwable t) {
+               throw new CacheException(t);
+            }
+         });
+         // We want to time out the current command future, not the main topology-waiting future,
+         // but the command future can take longer time to finish.
+         // This scheduled future does not have to be cancelled because it won't do anything (it will
+         // sit idle in the scheduler queue).
```

Comment:
Based on the comments in `ScheduledThreadPoolExecutor.DelayedWorkQueue`, all operations are `O(log n)`. Running a task is also `O(log n)`, because it pushing one task to the front and restoring the heap invariant. But the constant factor is larger, because it involves waking up the timer thread.